Engine/DataFeeds/Enumerators/FillForwardEnumerator.cs

```diff
@@ -288,7 +290,8 @@ protected virtual bool RequiresFillForwardData(TimeSpan fillForwardResolution, B
                 var potentialUtc = _offsetProvider.ConvertToUtc(item.ReferenceDateTime) + item.Interval;
                 var potentialInTimeZone = _offsetProvider.ConvertFromUtc(potentialUtc);
                 var potentialBarEndTime = RoundDown(potentialInTimeZone, item.Interval);
-                if (potentialBarEndTime < next.EndTime)
+                var nextEndTime = _offsetProvider.ConvertFromUtc(previousTimeUtc) + nextPreviousTimeDelta + _dataResolution;
+                if (potentialBarEndTime < nextEndTime)
```

Comment:
Changes in this method are high risk and hard to understand 😅 , I believe we need more code comments about what's happening here for example: why `_offsetProvider.ConvertFromUtc(previousTimeUtc)` and not directly `previous.Time`? Why can't we use `next.EndTime` directly? Why `previous.Time + nextPreviousTimeDelta` instead of `previous.EndTime`? Else we will forget and change it back. Think we need simple unit tests covering these `why` too.